modules/nix/linux-builder.nix

```diff
@@ -0,0 +1,78 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  inherit (pkgs) stdenv;
+
+  cfg = config.nix.linux-builder;
+in
+
+{
+  options.nix.linux-builder = {
+    enable = mkEnableOption "Linux builder";
+    package = mkOption {
+      type = types.package;
+      default = pkgs.darwin.builder;
+      defaultText = "pkgs.darwin.builder";
+      example = literalExpression "pkgs.darwin.builder.override { modules = [ { virtualisation.darwin-builder.hostPort = 22; } ] }";
+      description = ''
+        This option specifies the Linux builder to use.
+
+        You can add extra NixOS modules and configuration to the builder by overriding `modules`. You should be
+        running the Linux builder without any overrides (by not setting this option) before changing this option.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    assertions = [ {
+      assertion = config.nix.settings.trusted-users != [ "root" ] || config.nix.settings.extra-trusted-users != [ ];
+      message = ''
+        Your user or group (@admin) needs to be added to `nix.settings.trusted-users` or `nix.settings.extra-trusted-users`
+        to use the Linux builder.
+      '';
+    } ];
+
+    system.activationScripts.preActivation.text = ''
+      mkdir -p /var/lib/darwin-builder
+    '';
+
+    launchd.daemons.linux-builder = {
+      environment = {
+        inherit (config.environment.variables) NIX_SSL_CERT_FILE;
+      };
+      serviceConfig = {
+        ProgramArguments = [
+          "/bin/sh" "-c"
+          "/bin/wait4path /nix/store &amp;&amp; exec ${cfg.package}/bin/create-builder"
+        ];
+        KeepAlive = true;
+        RunAtLoad = true;
+        WorkingDirectory = "/var/lib/darwin-builder";
+        StandardOutPath = "/var/log/darwin-builder.log";
+        StandardErrorPath = "/var/log/darwin-builder.log";
+      };
+    };
+
+    environment.etc."ssh/ssh_config.d/100-linux-builder.conf".text = ''
+      Host linux-builder
+        Hostname localhost
+        HostKeyAlias linux-builder
+        Port 31022
+    '';
+
+    nix.distributedBuilds = true;
+
+    nix.buildMachines = [{
+      hostName = "linux-builder";
+      sshUser = "builder";
+      sshKey = "/etc/nix/builder_ed25519";
+      system = "${stdenv.hostPlatform.uname.processor}-linux";
+      supportedFeatures = [ "kvm" "benchmark" "big-parallel" ];
+      publicHostKey = "c3NoLWVkMjU1MTkgQUFBQUMzTnphQzFsWkRJMU5URTVBQUFBSUpCV2N4Yi9CbGFxdDFhdU90RStGOFFVV3JVb3RpQzVxQkorVXVFV2RWQ2Igcm9vdEBuaXhvcwo=";
```

Comment:
Is this completely static? Would be better to retrieve this information from an attribute of the builder package if possible.

Reply:
The host key is currently completely static, the host keys are files in the Nixpkgs tree and are not easily overridable, this value comes from the Nixpkgs documentation:

https://github.com/NixOS/nixpkgs/blob/master/doc/builders/special/darwin-builder.section.md?plain=1#L54